File: WinPhone/SmartCampusLibrary/indoormodel/graph/DictionaryGraph.cs
/*
Copyright (c) 2014, Aalborg University
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of the <organization> nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL AAlBORG UNIVERSITY BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace com.smartcampus.indoormodel.graph
{
	public class DictionaryGraph : IGraph {
	//******** Properties ******************//
	//data structure to look up vertex by id
	protected Dictionary<int, Vertex> vertices = new Dictionary<int, Vertex>();
	//data structure to get vertices by floor
	protected Dictionary<int, List<Vertex>> verticesByFloor = new Dictionary<int, List<Vertex>>();
	
	protected List<Edge> edges = new List<Edge>();
	
	protected List<Vertex> starcaseVertices = new List<Vertex>();
	protected List<Vertex> elevatorVertices = new List<Vertex>();

	public Edge addDirectionalEdge(Edge e) {
		Vertex origin = e.getOrigin();
		Vertex destination = e.getDestination();
		origin.addOutEdge(e); //destination
		destination.addInEdge(e); //origin
		
		//Add, if missing vertices
		if (!vertices.ContainsKey(origin.getId()))
			vertices.Add(origin.getId(), origin);
		
		if (!vertices.ContainsKey(destination.getId()))
			vertices.Add(destination.getId(), destination);
		
		if (!edges.Contains(e))
			edges.Add(e);
		
		return e;
	}
	
	//We also add any missing vertices
	public Edge addDirectionalEdge(Vertex origin, Vertex destination)
	{
		Edge e = new Edge(origin, destination);  
		return addDirectionalEdge(e);
	}
	
	public bool addElevatorVertex(Vertex v)
	{
		elevatorVertices.Add(v);
		return true;
	}
	
	public bool addStaircaseVertex(Vertex v)
	{
		starcaseVertices.Add(v);
		return true;
	}
	
	public Edge addUndirectionalEdge(Edge e) {
		Vertex origin = e.getOrigin();
		Vertex destination = e.getDestination();
		origin.addOutEdge(e); 
		destination.addOutEdge(e);
		origin.addInEdge(e);
		destination.addInEdge(e); 
		
		//Add, if missing vertices
		if (!vertices.ContainsKey(origin.getId()))
			vertices.Add(origin.getId(), origin);
		
		if (!vertices.ContainsKey(destination.getId()))
			vertices.Add(destination.getId(), destination);
		
		if (!edges.Contains(e))
			edges.Add(e);
		
		return e;
	}
	
	public IEnumerable<Edge> addUndirectionalEdges(Vertex v1, Vertex v2)
	{
		Edge e1 = addDirectionalEdge(v1, v2);
		Edge e2 = addDirectionalEdge(v2, v1);
		List<Edge> result = new List<Edge>();
		result.Add(e1);
		result.Add(e2);
		return result;
	}
	
	public bool addVertex(Vertex v)
	{
		if (vertices.ContainsKey(v.getId()))
			return false;
					
		//Add vertex to <vertexId, vertex> structure
		vertices.Add(v.getId(), v);
		//Add vertex to <floorNum, List<Vertex> structure
		if (v.getLocation() != null && v.getLocation().getAbsoluteLocation() != null)
		{
			int floor = (int)v.getLocation().getAbsoluteLocation().getAltitude();
			if (!verticesByFloor.ContainsKey(floor))
			{
				verticesByFloor.Add(floor, new List<Vertex>());
			}
			verticesByFloor[floor].Add(v);
		}
		return true;
	}
	
	public IEnumerable<Vertex> adjacentVertices(Vertex v)
	{
		return v.adjacentVertices();
	}
	
	public bool areAdjacent(Vertex v, Vertex w)
	{
		foreach (Vertex cur in v.adjacentVertices())
			if (cur.Equals(w))
				return true;
		return false;
	}
	
	public bool ContainsEdge(Edge e)
	{
		return edges.Contains(e);
	}
	
	public bool ContainsVertex(Vertex v)
	{
		return vertices.ContainsKey(v.getId());
	}
	
	public int degree(Vertex v)
	{
		if (vertices.ContainsValue(v))
			return v.degree();
		else
			return -1;
	}
	
	public IEnumerable<Vertex> destinations(Vertex v)
	{
		return v.destinations();
	}

	public Vertex[] endVertices(Edge e)
	{
		Vertex[] result = new Vertex[2];
		result[0] = e.getOrigin();
		result[1] = e.getDestination();
		return result;
	}
	
	public Vertex getClosestVertex(AbsoluteLocation userAbsLoc) {
		/*
		 * this may not be the best way of determining
		 * the vertex closest to the geopoint - a better solution
		 * could be to arrange the vertices in a different data structure
		 * (currently it is a list)
		 */
		
		int floor = (int)userAbsLoc.getAltitude();
		double dist, bestDist = double.MaxValue;
		Vertex closestVertex = null;
		foreach (Vertex v in this.getVertices(floor)) {
			AbsoluteLocation loc = v.getLocation().getAbsoluteLocation();
			dist = getDistance(loc.getLatitude() * 1E6, 
							   userAbsLoc.getLatitude() * 1E6, 
							   loc.getLongitude() * 1E6, 
							   userAbsLoc.getLongitude() * 1E6);
			
			if(dist < bestDist) {
				bestDist = dist;
				closestVertex = v;
			}
		}
		return closestVertex;
	}
	
	private double getDistance(double x1, double x2, double y1, double y2) {
		return Math.Sqrt(Math.Pow(x1 - x2, 2) + Math.Pow(y1 - y2, 2));
	}
	
	public List<Edge> getEdges()
	{
		return edges;
	}
	
	public List<Edge> getEdges(int floor)
	{
		List<Edge> result = new List<Edge>();
		foreach (Edge e in edges)
		if (e.getOrigin().getLocation().getAbsoluteLocation().getAltitude() == floor ||
			e.getDestination().getLocation().getAbsoluteLocation().getAltitude() == floor)
		{
			result.Add(e);
		}
		return result;
	}
	
	public List<Vertex> getElevatorVertices() {
		return elevatorVertices;
	}
	
	public List<Vertex> getElevatorVertices(int level) {
		List<Vertex> result = new List<Vertex>();
		foreach (Vertex v in elevatorVertices) {
			int vFloor = (int)(v.getLocation().getAbsoluteLocation().getAltitude());
			if (vFloor == level)
				result.Add(v);    		
		}
		return result;
	}
	
	public List<Vertex> getStaircaseVertices() {
		return starcaseVertices;
	}
	
	public List<Vertex> getStaircaseVertices(int level) {
		List<Vertex> result = new List<Vertex>();
		foreach (Vertex v in starcaseVertices) {
			int vFloor = (int)(v.getLocation().getAbsoluteLocation().getAltitude());
			if (vFloor == level)
				result.Add(v);
		}
		return result;
	}
	
	public Vertex getVertexById(int vertexId)
	{
		return vertices[vertexId];
	}

	public IEnumerable<Vertex> getVertices()
	{
        return vertices.Values;        
	}
			
	public List<Vertex> getVertices(int floor)
	{
		if (!verticesByFloor.ContainsKey(floor))
			return null;
		else
			return verticesByFloor[floor]; //may also be null	
	}   
	
	public IEnumerable<Edge> incidentEdges(Vertex v)
	{
		return v.incidentEdges();
	}
	
	public int inDegree(Vertex v)
	{
		return v.inDegree();
	}    
		
	public IEnumerable<Edge> inEdges(Vertex v)
	{
		return v.getInEdges();
	}
	
	// Methods dealing with directed edges //

	//Methods dealing with positioning 
	//We don't bother to create a subclass for this behavior as our graph is only used for one purpose in this application
	public void InsertRadiusVertices(Vertex v, int radius)
	{
		AbsoluteLocation sourceLocation = v.getLocation().getAbsoluteLocation(); 
		AbsoluteLocation targetLocation;
		double dist;
		foreach (Vertex w in vertices.Values)
		{
			if (v.Equals(w))
				continue;
			
			targetLocation = w.getLocation().getAbsoluteLocation(); 
			dist = com.smartcampus.baselogic.DistanceMeasurements.CalculateMoveddistanceInMeters(
					sourceLocation.getLatitude(), sourceLocation.getLongitude(),
					targetLocation.getLatitude(), targetLocation.getLongitude());
			
			if (dist <= radius)
				v.addRadiusVertex((Vertex)w);
		}
	}
	
	public int numEdges()
	{
		return edges.Count;
	}
	
	public int numVertices()
	{
		return vertices.Count;
	}
	
	public Vertex opposite(Edge e, Vertex v)
	{
		return e.Opposite(v);
	}
	
	/**
	 * @return All vertices with an edge going to v
	 */
	public IEnumerable<Vertex> origins(Vertex v)
	{
		return v.origins();
	}
	
	public int outDegree(Vertex v)
	{
		return v.outDegree();
	}        
	
	public IEnumerable<Edge> outEdges(Vertex v)
	{
		return v.getOutEdges();
	}
	
	public bool removeDirectionalEdge(Vertex origin, Vertex destination)
	{
		Edge e = new Edge(origin, destination);    
		//Removes edge if present
		bool sourceMod = 
			origin.removeOutEdge(destination) ||
			destination.removeInEdge(origin);
		return edges.Remove(e) || sourceMod;  
	}
	
	public bool removeElevatorVertex(Vertex v)
	{
		return elevatorVertices.Remove(v);
	}
	
	public bool removeStaircaseVertex(Vertex v)
	{
		return starcaseVertices.Remove(v);
	}
	
	public bool removeUndirectionalEdges(Vertex v1, Vertex v2)
	{
		bool mod1 = removeDirectionalEdge(v1, v2);
		bool mod2 = removeDirectionalEdge(v2, v1);
		return mod1 || mod2;
	}

	public bool removeVertex(Vertex v)
	{
		foreach (Edge e in v.incidentEdges())
			edges.Remove(e);
			
		//remove from <vertexId, vertex> structure
		vertices.Remove(v.getId());
		//remove from <floorNum, List<Vertex> structure
		bool hasLocation = v.getLocation() != null && v.getLocation().getAbsoluteLocation() != null;
		if (hasLocation)
		{
			int floor = (int)v.getLocation().getAbsoluteLocation().getAltitude();
			if (verticesByFloor.ContainsKey(floor))
			{
				verticesByFloor[floor].Remove(v);
			}
		}
		return true;
	}
}
}

File: iOS/SmartCampusApp/SmartCampus/DictionaryGraph.m
//
//  DictionaryGraph.m
//  SmartCampus
//
// Copyright (c) 2014, Aalborg University
// All rights reserved.

// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//     * Neither the name of the <organization> nor the
//       names of its contributors may be used to endorse or promote products
//       derived from this software without specific prior written permission.

// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL AAlBORG UNIVERSITY BE LIABLE FOR ANY
// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import "DictionaryGraph.h"
#import "DistanceMeasurements.h"
#import "Constants.h"

@implementation DictionaryGraph

+ (DictionaryGraph *)graphWithJson:(NSDictionary *)data {
    DictionaryGraph *graph = [DictionaryGraph new];

    for (id vertexData in [data valueForKey:@"Vertices"]) {
        [graph addVertex:[Vertex vertexWithJson:vertexData]];
    }

    for (id edgeData in [data valueForKey:@"Edges"]) {
        Vertex *origin = [graph vertexForId:[[edgeData valueForKey:@"vertexOrigin"] intValue]];
        Vertex *destination = [graph vertexForId:[[edgeData valueForKey:@"vertexDestination"] intValue]];

        for (Edge *edge in [graph addUndirectionalEdgesBetweenVertex:origin andVertex:destination]) {
            edge.id = [[edgeData valueForKey:@"ID"] intValue];
            edge.buildingId = [[edgeData valueForKey:@"Building_ID"] intValue];
            edge.staircase = [[edgeData valueForKey:@"is_stair"] boolValue];
            edge.elevator = [[edgeData valueForKey:@"is_elevator"] boolValue];
        }
    }

    return graph;
}

- (id)init
{
    if(!(self = [super init]))
        return nil;

    self.vertices = [NSMutableDictionary new];
    self.verticesByFloor = [NSMutableDictionary new];
    self.edges = [NSMutableArray new];
    self.staircaseEndpoints = [NSMutableArray new];
    self.elevatorEndpoints = [NSMutableArray new];

    return self;
}

- (Edge *)addDirectionalEdge:(Edge *)edge {
    Vertex *origin = edge.origin;
    Vertex *destination = edge.destination;

    [origin addOutEdge:edge];
    [destination addInEdge:edge];

    [self addVertex:origin];
    [self addVertex:destination];

    if (![self.edges containsObject:edge]) {
        [self.edges addObject:edge];
    }

    return edge;
}

- (Edge *)addDirectionalEdgeFromOrigin:(Vertex *)origin toDestination:(Vertex *)destination {
    Edge *edge = [[Edge alloc] initWithOrigin:origin andDestination:destination];
    return [self addDirectionalEdge:edge];
}

- (BOOL)addElevatorVertex:(Vertex *)vertex {
    [self.elevatorEndpoints addObject:vertex];
    return YES;
}

- (BOOL)addStaircaseVertex:(Vertex *)vertex {
    [self.staircaseEndpoints addObject:vertex];
    return YES;
}

- (Edge *)addUndirectionalEdge:(Edge *)edge {
    Vertex *origin = edge.origin;
    Vertex *destination = edge.destination;

    [origin addOutEdge:edge];
    [destination addOutEdge:edge];
    [origin addInEdge:edge];
    [destination addInEdge:edge];

    [self addVertex:origin];
    [self addVertex:destination];

    if(![self.edges containsObject:edge])
        [self.edges addObject:edge];

    return edge;
}

- (NSArray *)addUndirectionalEdgesBetweenVertex:(Vertex *)vertex1 andVertex:(Vertex *)vertex2 {
    Edge *edge1 = [self addDirectionalEdgeFromOrigin:vertex1 toDestination:vertex2];
    Edge *edge2 = [self addDirectionalEdgeFromOrigin:vertex2 toDestination:vertex1];

    return [[NSArray alloc] initWithObjects:edge1, edge2, nil];
}

- (BOOL)addVertex:(Vertex *)vertex {
    if ([self containsVertex:vertex]) return NO;
    
    [self.vertices setObject:vertex forKey:[NSNumber numberWithInt:vertex.id]];
    
    if (vertex.location && vertex.location.absolute) {
        NSNumber *floor = [NSNumber numberWithInt:(int) vertex.location.absolute.altitude];

        if (![self.verticesByFloor objectForKey:floor]) {
            [self.verticesByFloor setObject:[NSMutableArray new] forKey:floor];
        }

        [(NSMutableArray *) [self.verticesByFloor objectForKey:floor] addObject:vertex];
    }
    
    return YES;
}

- (NSArray *)verticesAdjacentToVertex:(Vertex *)vertex {
    return [vertex adjacentVertices];
}

- (BOOL)isVertex:(Vertex *)vertex1 adjacentTo:(Vertex *)vertex2 {
    for (Vertex *vertex in [vertex1 adjacentVertices]) {
        if ([vertex isEqual:vertex2]) return YES;
    }

    return NO;
}

- (BOOL)containsEdge:(Edge *)edge {
    return [self.edges containsObject:edge];
}

- (BOOL)containsVertex:(Vertex *)vertex {
    return [self vertexForId:vertex.id] != nil;
}

- (int)degreeOfVertex:(Vertex *)vertex {
    return [self containsVertex:vertex] ? [vertex degree] : -1;
}

- (NSArray *)destinationsOfVertex:(Vertex *)vertex {
    return [vertex destinations];
}

- (NSArray *)endVerticesOfEdge:(Edge *)edge {
    return [[NSArray alloc] initWithObjects:edge.origin, edge.destination, nil];
}

- (Vertex *)closestVertexToLocation:(AbsoluteLocation *)location {
    int floor = (int) location.altitude;

    double dist, bestDist = DBL_MAX;
    Vertex *closestVertex = nil;

    for (Vertex *vertex in [self verticesForFloor:floor]) {
        dist = [self getDistance
                :[vertex latitude] * 1E6 :location.latitude *1E6
                :[vertex longitude] *1E6 :location.longitude *1E6];

        if(dist < bestDist){
            bestDist = dist;
            closestVertex = vertex;
        }
    }

    return closestVertex;
}

//private method
- (double)getDistance:(double)x1: (double)x2: (double)y1: (double)y2 
{
    return sqrt(pow(x1 - x2, 2) + pow(x1 - x2, 2));
}

- (NSArray *)edgesForFloor:(int)floor {
    NSMutableArray *edges = [NSMutableArray new];

    for (Edge *edge in self.edges) {
        if (edge.origin.altitude == floor || edge.destination.altitude == floor) {
            [edges addObject:edge];
        }
    }

    return edges;
}

- (NSArray *)elevatorEndpointsForFloor:(int)floor {
    NSMutableArray *endpoints = [NSMutableArray new];
    
    for (Vertex *vertex in self.elevatorEndpoints) {
        if (((int) vertex.altitude) == floor) [endpoints addObject:vertex];
    }

    return endpoints;
}

- (NSArray *)staircaseEndpointsForFloor:(int)floor {
    NSMutableArray *endpoints = [NSMutableArray new];

    for (Vertex *vertex in self.staircaseEndpoints) {
        if (((int) vertex.altitude) == floor) [endpoints addObject:vertex];
    }

    return endpoints;
}

- (Vertex *)vertexForId:(int)id {
    return [self.vertices objectForKey:[NSNumber numberWithInt:id]];
}

- (NSArray *)verticesForFloor:(int)floor {
    return [self.verticesByFloor objectForKey:[NSNumber numberWithInt:floor]];
}

- (void)insertRadiusVerticesAtVertex:(Vertex *)source withRadius:(int)radius {
    double dist;
    for(Vertex *target in self.vertices) {
        if ([source isEqual:target]) continue;

        dist = [DistanceMeasurements distanceInMetersFromVertex:source toVertex:target];

        if (dist <= radius) [source addRadiusVertex:target];
    }
}

- (int)numEdges
{
    return [self.edges count];
}

- (int)numVertices
{
    return [self.vertices count];
}

- (BOOL)removeDirectionalEdgeFromOrigin:(Vertex *)origin toDestination:(Vertex *)destination {
    Edge *edge = [[Edge alloc] initWithOrigin:origin andDestination:destination];

    BOOL sourceModified = [origin removeOutEdge:destination] || [destination removeInEdge:origin];
    BOOL edgeModified = NO;
    if ([self.edges containsObject:edge]) {
        [self.edges removeObject:edge];
        edgeModified = YES;
    }

    return sourceModified || edgeModified;
}

- (BOOL)removeElevatorEndpoint:(Vertex *)endpoint {
    if([self.elevatorEndpoints containsObject:endpoint]) {
        [self.elevatorEndpoints removeObject:endpoint];

        return YES;
    }

    return NO;
}

- (BOOL)removeStaircaseEndpoint:(Vertex *)endpoint {
    if([self.staircaseEndpoints containsObject:endpoint]) {
        [self.staircaseEndpoints removeObject:endpoint];

        return YES;
    }

    return NO;
}

- (BOOL)removeUndirectionalEdges:(Vertex *)vertex1: (Vertex *)vertex2 {
    return [self removeDirectionalEdgeFromOrigin:vertex1 toDestination:vertex2]
    || [self removeDirectionalEdgeFromOrigin:vertex2 toDestination:vertex1];
}

- (BOOL)removeVertex:(Vertex *)vertex {
    for (Edge *edge in vertex.incidentEdges) {
        [self.edges removeObject:edge];
    }
    
    [self.vertices removeObjectForKey:[NSNumber numberWithInt:vertex.id]];
    
    if (vertex.location != nil && vertex.location.absolute != nil) {
        int floor = (int) vertex.altitude;
        
        NSMutableArray *verticesOnFloor = [self.verticesByFloor objectForKey:[NSNumber numberWithInt:floor]];
        if (verticesOnFloor) [verticesOnFloor removeObject:vertex];
    }

    return YES;
}

- (NSString *)description {
    NSMutableString* desc = [NSMutableString stringWithFormat:@"Graph with %d vertices and %d edges:",
                             self.numVertices, self.numEdges];

    [desc appendString:@"\r  Vertices:"];
    for (id vertex in [self.vertices allValues]) {
        [desc appendFormat:@"\r    %@", vertex];
    }

    [desc appendString:@"\r  Edges:"];
    for (id edge in self.edges) {
        [desc appendFormat:@"\r    %@", edge];
    }

    return desc;
}

@end
